src/3d/symbols/qgspointcloud3dsymbol.cpp

```diff
@@ -485,28 +572,38 @@ QgsColorRampShader QgsClassificationPointCloud3DSymbol::colorRampShader() const
-  Qt3DRender::QParameter *renderingStyle = new Qt3DRender::QParameter( "u_renderingStyle", QgsPointCloud3DSymbol::ColorRamp );
-  mat->addParameter( renderingStyle );
   Qt3DRender::QParameter *pointSizeParameter = new Qt3DRender::QParameter( "u_pointSize", QVariant::fromValue( mPointSize ) );
   mat->addParameter( pointSizeParameter );
-  // Create the texture to pass the color ramp
-  Qt3DRender::QTexture1D *colorRampTexture = nullptr;
-  if ( mColorRampShader.colorRampItemList().count() > 0 )
+
+  if ( !mRenderAsTriangles )
   {
-    colorRampTexture = new Qt3DRender::QTexture1D( mat );
-    colorRampTexture->addTextureImage( new QgsColorRampTexture( mColorRampShader, 1 ) );
-    colorRampTexture->setMinificationFilter( Qt3DRender::QTexture1D::Linear );
-    colorRampTexture->setMagnificationFilter( Qt3DRender::QTexture1D::Linear );
-  }
+    // Create the texture to pass the color ramp
+    const QgsColorRampShader mColorRampShader = colorRampShader();
+    Qt3DRender::QParameter *renderingStyle = new Qt3DRender::QParameter( "u_renderingStyle", QgsPointCloud3DSymbol::ColorRamp );
+    mat->addParameter( renderingStyle );
+    Qt3DRender::QTexture1D *colorRampTexture = nullptr;
+    if ( mColorRampShader.colorRampItemList().count() > 0 )
+    {
+      colorRampTexture = new Qt3DRender::QTexture1D( mat );
+      colorRampTexture->addTextureImage( new QgsColorRampTexture( mColorRampShader, 1 ) );
+      colorRampTexture->setMinificationFilter( Qt3DRender::QTexture1D::Linear );
+      colorRampTexture->setMagnificationFilter( Qt3DRender::QTexture1D::Linear );
+    }
 
-  // Parameters
-  Qt3DRender::QParameter *colorRampTextureParameter = new Qt3DRender::QParameter( "u_colorRampTexture", colorRampTexture );
-  mat->addParameter( colorRampTextureParameter );
-  Qt3DRender::QParameter *colorRampCountParameter = new Qt3DRender::QParameter( "u_colorRampCount", mColorRampShader.colorRampItemList().count() );
-  mat->addParameter( colorRampCountParameter );
-  const int colorRampType = mColorRampShader.colorRampType();
-  Qt3DRender::QParameter *colorRampTypeParameter = new Qt3DRender::QParameter( "u_colorRampType", colorRampType );
-  mat->addParameter( colorRampTypeParameter );
+    // Parameters
+    Qt3DRender::QParameter *colorRampTextureParameter = new Qt3DRender::QParameter( "u_colorRampTexture", colorRampTexture );
+    mat->addParameter( colorRampTextureParameter );
+    Qt3DRender::QParameter *colorRampCountParameter = new Qt3DRender::QParameter( "u_colorRampCount", mColorRampShader.colorRampItemList().count() );
+    mat->addParameter( colorRampCountParameter );
+    const int colorRampType = mColorRampShader.colorRampType();
+    Qt3DRender::QParameter *colorRampTypeParameter = new Qt3DRender::QParameter( "u_colorRampType", colorRampType );
+    mat->addParameter( colorRampTypeParameter );
+  }
+  else
+  {
+    // When we triangulate, the color is passed to the shader instead of the parameter
```

Comment:
Why do we need to pass the color to the shader instead of letting shader to calculate it?

Reply:
With triangles, the shader interpolates between values associated to each vertices to render the triangle. With color ramp, we sent the parameter to the shader, the parameter is interpolated for each fragment of triangles and the color ramp is applied to obtain the good color of the fragment.
With classification, we need to render the color corresponding exactly to the parameter value, so the above principle doesn't work anymore. Indeed, as the parameter value are interpolated between vertices, no exact value anymore except if vertices of a triangle have the same value. Applying the color ramp in the shader leads to black triangles it they don't have the same value on vertices, so we obtain a bad rendering. To avoid this, we need to calculate the color before sending it to the shader, then the shader interpolates the color on triangle, not the parameter, and the rendering is good.

Reply:
I see. How about using `flat` interpolation qualifier for the parameter, which would disable interpolation within a triangle and use a single value from one vertex for the whole triangle? https://www.khronos.org/opengl/wiki/Type_Qualifier_(GLSL)#Interpolation_qualifiers

I have not tried it myself, but it could help. I am just trying to avoid adding extra code paths that may complicate the code.

Alternatively, I think from what I read is that if we pass class number as an integer in GLSL, it will not be interpolated, so maybe the float "parameter" could be used for other renderer types and a "classIndex" just for classified renderer?

Reply:
Indeed, using `flat` seems to be the solution with adding a new directions in the vertex shader depending of the type of classification.